File: backend/hackathon-be/src/main/java/com/example/hackathonbe/common/exceptions/dto/ApiError.java
package com.example.hackathonbe.common.exceptions.dto;

import java.time.OffsetDateTime;
import java.util.List;

public record ApiError(
        String code,
        String message,
        int status,
        String path,
        OffsetDateTime timestamp,
        List<FieldViolation> fieldErrors
) {
    public record FieldViolation(String field, String message, Object rejectedValue) {}
}

File: frontend/hackathon/src/types/apiError.ts
export type ApiError = {
    code?: string
    message?: string
    status?: number
    path?: string
    timestamp?: string
    fieldErrors?: Array<{ field: string; message: string }> | null
}

export async function readApiError(res: Response): Promise<string> {
    const contentType = res.headers.get('content-type') ?? ''

    // Prefer JSON
    if (contentType.includes('application/json')) {
        try {
            const data = (await res.json()) as ApiError
            if (data?.message) return data.message
        } catch {
            // fall through
        }
    }

    // Fallback to text
    const text = await res.text().catch(() => '')
    if (!text) return `Request failed (status ${res.status})`

    // Defensive JSON parse
    try {
        const parsed = JSON.parse(text) as ApiError
        if (parsed?.message) return parsed.message
    } catch {}

    return text
}
